pkg/gatewayserver/redis/registry.go

```diff
@@ -34,15 +36,16 @@ func (r *GatewayConnectionStatsRegistry) key(uid string) string {
 
 // Set sets or clears the connection stats for a gateway.
 func (r *GatewayConnectionStatsRegistry) Set(ctx context.Context, ids ttnpb.GatewayIdentifiers, stats *ttnpb.GatewayConnectionStats) error {
-	uid := unique.ID(ctx, ids)
-
+	key := r.key(unique.ID(ctx, ids))
 	defer trace.StartRegion(ctx, "set gateway connection stats").End()
 
 	var err error
 	if stats == nil {
-		err = r.Redis.Del(r.key(uid)).Err()
+		err = r.Redis.Del(key).Err()
+		r.allKeys.Delete(key)
 	} else {
-		_, err = ttnredis.SetProto(r.Redis, r.key(uid), stats, 0)
+		_, err = ttnredis.SetProto(r.Redis, key, stats, 0)
+		r.allKeys.Store(key, struct{}{})
```

Comment:
If we only store, and if the gateway reconnects to a different instance, and the current instance goes down, the gateway connection stats get reset. That shouldn't happen; if the gateway disconnects, it should be cleared from the map.

Also, have you considered alternative ways to keep a unique set of strings?

How about storing them in a set in Redis per GS instance?

Reply:
> If we only store, and if the gateway reconnects to a different instance, and the current instance goes down, the gateway connection stats get reset. That shouldn't happen; if the gateway disconnects, it should be cleared from the map.

When the gateway disconnects from this instance, the stats will be cleared anyway (while the `io.Connection` is shutting down) which would also remove the key from the map.

> Also, have you considered alternative ways to keep a unique set of strings?
>
> How about storing them in a set in Redis per GS instance?

This does feel like a better solution, and we can change `Set` to update in a transaction, so we would be largedly avoiding the issue above.

If we do have a unique instance id then this could also be a solution. I would like that ID to persist on restarts as well, but that is not really needed.